React/Base/RCTJavaScriptLoader.mm

```diff
@@ -142,10 +142,19 @@ + (NSData *)attemptSynchronousLoadOfBundleAtURL:(NSURL *)scriptURL
     return nil;
 
   case facebook::react::ScriptTag::BCBundle:
-    if (header.BCVersion != runtimeBCVersion) {
+    if (runtimeBCVersion == JSNoBytecodeFileFormatVersion || runtimeBCVersion < 0) {
+      if (error) {
+        *error = [NSError errorWithDomain:RCTJavaScriptLoaderErrorDomain
+                                     code:RCTJavaScriptLoaderErrorBCNotSupported
+                                 userInfo:@{NSLocalizedDescriptionKey:
+                                              @"Bytecode bundles are not supported by this runtime."}];
+      }
+      return nil;
+    }
+    else if ((uint32_t)runtimeBCVersion != header.BCVersion) {
       if (error) {
         NSString *errDesc =
-          [NSString stringWithFormat:@"BC Version Mismatch. Expect: %d, Actual: %d",
+          [NSString stringWithFormat:@"BC Version Mismatch. Expect: %d, Actual: %u",
                     runtimeBCVersion, header.BCVersion];
```

Comment:
`header.BCVersion` is pulled directly out of the byte code header and it's unsigned, which obviously makes sense. `runtimeBCVersion` is a signed argument allowing `JSNoBytecodeFileFormatVersion = -1` to be used as a magic value indicating no support (although in this logic I've assumed _any_ negative value or equality with that magic value indicates runtime incompatibility).

The idea is we determine that `runtimeBCVersion` is non-negative (i.e. MSB=0) before comparing it to the unsigned `header.BCVersion` - it's safe to cast the former to `uint32_t` at that point and do a comparison.